System/Random.hs

```diff
@@ -40,6 +40,72 @@
 --   instance of 'Random' allows one to generate random values of type
 --   'Float'.
 --
+-- [/Example for RNG Implementors:/]
+--
+-- For example, we can define a [linear congruential
+-- generator](https://en.wikipedia.org/wiki/Linear_congruential_generator):
+--
+-- >>> let lcg = (\n -> (n * 1103515245 + 12345) `mod` 2^31) :: Word32 -> Word32
+-- >>> lcg 42
+-- 1250496027
+--
+-- And then make it an instance of `RandomGen`:
+--
+-- >>> data LcgGen = LcgGen Word32
+--
+-- >>> :{
+-- instance RandomGen LcgGen where
+--   next (LcgGen w) = (fromIntegral x, LcgGen x)
+--     where
+--       x = lcg w
+--   split (LcgGen w) = (LcgGen x, LcgGen y)
+--     where
+--       x = lcg w
+--       y = lcg x
+-- :}
+--
+-- >>> :{
+-- let randomListM :: (Random a, MonadRandom g m, Num a, Uniform a) => g -> Int -> m [a]
+--     randomListM gen n = replicateM n (uniform gen)
+-- :}
+--
+-- >>> :{
+-- let rolls :: [Word32]
+--     rolls = runStateGen_
+--               (LcgGen 1729 :: LcgGen)
+--               (randomListM PureGen 10 >>= \xs -> return $ map ((+1) . (`mod` 6)) xs)
+-- :}
+--
+-- >>> rolls
+-- [1,6,5,2,3,2,5,2,3,2]
+--
+-- Importantly, this implementation will not be as efficient as it
+-- could be because the random values are converted to 'Integer' and
+-- then to desired type.
+--
+-- Instead we should define (where @unBuildWord32 :: Word32 ->
+-- (Word16, Word16)@ is a function to pull apart a 'Word32' into a
+-- pair of 'Word16'):
+--
+-- >>> data LcgGen' = LcgGen' Word32
+--
+-- >>> :set -fno-warn-missing-methods
+--
+-- >>> :{
+-- instance RandomGen LcgGen' where
+--   genWord16 (LcgGen' w) = (y, LcgGen' x)
+--     where
+--      x = lcg w
+--      (y, _) = unBuildWord32 x
+--   genWord32 (LcgGen' w) = (x, LcgGen' x)
+--     where
+--       x = lcg w
```

Comment:
That's only 31 bits of randomness, not 32. I may be too pedantic with this example, but I would like to avoid people learning the wrong things from this API usage guide.